### PR TITLE
[FLINK-5971] [flip-6] Add timeout for registered jobs on the ResourceManager

### PR DESCRIPTION
**Pull request by tillrohrmann**
This PR introduces a timeout for inactive jobs on the ResourceManager. A job is inactive
if there is no active leader known for this job. In case that a job times out, it will
be removed from the ResourceManager. Additionally, this PR removes the dependency of
the JobLeaderIdService on the RunningJobsRegistry.

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.flink.annotation.PublicEvolving;
+
+/**
+ * The set of configuration options relating to the ResourceManager
+ */
+@PublicEvolving
+public class ResourceManagerOptions {
+
+	public static final ConfigOption<String> JOB_TIMEOUT = ConfigOptions
+		.key("resourcemanager.job.timeout")
```
Comment (KurtYoung) [inline flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java]: maybe we can change the name to something like "inactive_job.timeout" or "idle_job.timeout" to make this config more specific

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/JobLeaderIdService.java [hunk 2]
```diff
@@ -283,5 +302,31 @@ public void handleError(Exception exception) {
...
+			}
+
+			final UUID newTimeoutId = UUID.randomUUID();
+
+			timeoutId = newTimeoutId;
+
+			timeoutFuture = scheduledExecutor.schedule(new Runnable() {
+				@Override
+				public void run() {
+					listenerJobLeaderIdActions.notifyJobTimeout(jobId, newTimeoutId);
```
Comment (KurtYoung) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/JobLeaderIdService.java]: I think this thread is different with the leader retrieve thread which will call "notifyLeaderAddress", right?
So is this possible that when a job is close to timeout, and then comes the notification of leader address, both threads are trying to `cancelTimeout`, NPE will be thrown if one of them trying to call `timeoutFuture.cancel(true)`, while the other one has set `timeoutFuture = null`

File: flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java (hunk 1 above)
Comment (KurtYoung) [inline flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java]: can we have a more specific name like "inactive_job.timeout". It looks like normal job will also timeout for 5 minutes.

File: flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java [hunk 3]
```diff
@@ -28,6 +28,10 @@
 @PublicEvolving
 public class AkkaOptions {
 
+	public static final ConfigOption<String> AKKA_ASK_TIMEOUT = ConfigOptions
```
Comment (KurtYoung) [inline flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java]: Maybe add a comment to be consistency with other options in this class?

Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java]: True will add it.

File: flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java (hunk 1 above)
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java]: Hmm not sure whether it's clear to the user what an inactive job is. Will add java docs, though.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/JobLeaderIdService.java (hunk 2 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/JobLeaderIdService.java]: Yes I think you're right. Good point. Will fix the problem.

**Comment by tillrohrmann** [comment]
Thanks a lot for your review @KurtYoung. I've addressed you comments and updated the PR.

**Comment by KurtYoung** [comment]
+1 to merge.

**Comment by tillrohrmann** [comment]
Merging this PR.

**Pull request closed without merging**